core/src/main/java/org/apache/calcite/util/Sources.java

```diff
@@ -189,14 +196,25 @@ private static File urlToFile(URL url) {
       return Paths.get(uri).toFile();
     }
 
+    private static URL fileToUrl(File file) {
```

Comment:
I wonder if it should have something like 

```java
      String filePath = file.getPath();
      if (!file.isAbsolute()) {
        if (File.separatorChar != '/') {
          filePath = filePath.replace(File.separatorChar, '/');
        }
        if (file.isDirectory() && !filePath.endsWith("/")) {
          filePath += "/";
        }
        try {
          return new URL("file", null, 0, URLEncoder.encode(filePath, StandardCharsets.UTF_8.name()));
        } catch (MalformedURLException e) {
          throw new IllegalArgumentException("Unable to create URI for file " + filePath, e);
        } catch (UnsupportedEncodingException e) {
          throw new IllegalArgumentException("UTF-8 is not found", e);
        }
      }
```

Then implementation would handle both relative and absolute files symmetrically.
In other words, for the relative file it would produce something like `file:abc%20def.txt` which is in line with `FileSource(URL)`.

Reply:
@vlsi Thanks a lot for your feedback. I have revised the code accordingly.